swarm/network/stream/syncer.go

```diff
@@ -194,9 +194,9 @@ func RegisterSwarmSyncerClient(streamer *Registry, db *storage.DBAPI) {
 
 // NeedData
 func (s *SwarmSyncerClient) NeedData(key []byte) (wait func()) {
-	chunk, _ := s.db.GetOrCreateRequest(key)
+	chunk, need := s.db.GetOrCreateRequest(key)
 	// TODO: we may want to request from this peer anyway even if the request exists
-	if chunk.ReqC == nil {
+	if chunk.ReqC == nil || need == false {
```

Comment:
`!need` 

but why do we need this?

Reply:
From the `hq` chat:


When running tests, when it would supposedly shut down, I would still see a lot of "LocalStore.GetOrRetrieve: %v found locally" messages
The chain there is NeedData() in syncer.go:
```
197   chunk, need := s.db.GetOrCreateRequest(key)
198   // TODO: we may want to request from this peer anyway even if the request exists
199   if chunk.ReqC == nil {
200     return nil
201   }
202   // create request and wait until the chunk data arrives and is stored
203   return func() {
204     chunk.WaitToStore()
205   }
206 }
```
from `GetOrCreateRequest`, it always returns with `chunk.WaitToStore()` in these cases

Which is suspicious - why would it want to wait for the chunk data to arrive if it is already there?

so if I change line 199 to
`199   if chunk.ReqC == nil || need == false {`
all the tests pass
no deadlock

the other path I could see is that `chunk.ReqC `should be nil in that case, but I don't understand why